Implements Kindle screen rotate on emulator and forces portrait mode on doc close

kai771 opened the pull request:
Although I said “don’t look at me” in #573, it turned out to be pretty simple. This is not a full implementation - both J and K will just flip back and forth to landscape/portrait. I don’t think full implementation is necessary for development (who’d want to read upside down?), and I’m lazy by nature, so if anyone wants to do all modes - be my guest.

Second commit contains changes that will force portrait mode on closing documents. I tested it on both emu and Kindle, and it appears to work fine (at least better than without the patch. Try rotating the screen in a version prior to this PR, and then press Home. You’ll see what I mean). But, given my history of missing things, please do test yourself. I uploaded latest build [here](https://github.com/kai771/kindlepdfviewer/downloads), so it should be easy for anyone to test and report bugs.

kai771 committed:
Implements Kindle screen rotate on emulator

kai771 committed:
force portrait mode after closing document

dracodoc commented:
I'm going to test it to show my support... :P

dracodoc commented:
It seemed to me that keep pressing J will keep rotating just as before. Actually I'd prefer a user configurable landscape angle, then just using one key to switch between landscape and portrait. 

Right now I have to use different keys to switch between landscape and portrait: K for rotate, J for rotate back. If there is only one landscape angle, one key can switch it back and forth which is simpler and easy to remember.

The force portrait mode on closing is working fine. The next thing should be remember the rotating mode in history so next time it can be opened and rotated automatically.

kai771 commented:
@dracodoc Yeah, the only change I did on kindle is forcing portrait on document close. Function of J and K is not changed (on the Kindle. I implemented screen rotation for the emulator (it wasn't supported in the emulator before) and I was lazy to implement it fully, so in the emulator it just switches landscape/portrait, instead of really rotating it, so you can have upside down text. You don't normaly rotate your monitor, so I think it's not needed for emulation :) ). It's true that you need to use 2 keys to switch back and forth portrait/landscape, but they're right next to each other. Anyone can do it even with their eyes closed. They're "rotate" buttons. You want "portrait/landscape toggle". If we implemented it like that, I'm sure someone would complain that she can't rotate it (or that she has to choose different landscape angle before she gets what she wants, because she often switches hands holding the Kindle, and she prefers it with buttons on the left when holding it in landscape with the left hand, and buttons on the right when holding it with the right hand... let your imagination run wild). Supporting both is just too much effort for a simple little thing, that isn't at all difficult as it is now. Just my noob opinion... feel free to differ.

I have a similar opinion of the saving orientation with the book. Is it really that hard to press just 1 key after opening the book (and I thought I was lazy :D)? Well, I suppose I could see some justification for this... much more than for landscape/portrait toggle, at least.

(Note: sorry if this came out harsh, this is just my personal opinion, and I assure you I had no intention of offending anyone) 
Oh, right, and thank you for testing :).

dracodoc commented:
No problem. Since the current implementation is already there, it's probably not worth it to change to my preference.

Though I'd say saving orientation is definitely have many advantages, zero side effect and it should be really easy to do. If one book was to be read in landscape mode, user probably will read it later still in landscape mode. If user want to change to portrait mode anytime, he can press a key and change it.

kai771 commented:
I was more of an opinion that she would look at the TOC first, or bookmarks, or flip a few pages to see where she left off before she continues reading... all more easily done in portrait mode. What do you guys think? As for implementation, it should be easy, especially now that I implemented :setRotationMode() :).

dracodoc commented:
The last screen should at least have some content that user have read it before, so it should be enough to remind user. The bookmarks only provide some page number list. And he can always flip back in landscape mode to help remembering last read.

The key difference is, the landscape view is vastly different from the portrait view even for same page, same contents. So always keep in landscape view for one book is better to maintain the reading memory. 

At last, user can always press one key to return to portrait mode if he felt it is needed. So the question is, which case is more often, to read continuously in landscape mode or switch back to portrait, move around a bit then switch to landscape again?

kai771 commented:
I don't know about others, but I use fit-to-content-width almost all the time (when I don't, I use 2 column mode). Even if I used landscape, it would be fit-to-content-width. The page I see is the same in both portrait and landscape, and it looks the same - the text is just smaller in portrait, but there's more of it. So I wouldn't say it's vastly different. I must say I'd prefer if the page should be in portrait when I open it, because I already hold the kindle the right way. If I wanted to read in landscape, I'd already have to move my arms, so pressing a key on the way is not too much effort. Let's assess the following situation: you have many books, and you are not sure which one is the one you'd like to open. You'll open one book, see that that's not the one you're looking for, close it, open another... this is easier if you don't have to rotate the kindle after every opening to see if it's the book you were looking for.

And final reason - remembering book orientation requires more code. Always opening in portrait doesn't xD.

Again, these are just my opinions. The fact that I don't agree with you doesn't mean that others won't.

dracodoc commented:
The reason I say landscape mode is vastly different is that the amount of text you can see in one screen is very different. So with portrait mode you will have a memory of the whole page arrangement, but with landscape mode you probably can only remember the specific content and cannot see much page arrangement in that view.

Of course user can rotate the screen by pressing one key, but that will show the first part of current page, so it could be not the last reading position and user need to flip down several times to find the exact location.

The difference of our opinions are not much, we can always press one key to switch in either case, so it's just which mode should be default. Another solution will be make this behavior configurable, but that will introduce more complexity.

Anyway, this case is not a major one. I can live with the current status.

tigran123 commented:
Interesting, I will have a look at some point.

tigran123 commented:
At the moment I am testing only on the emulator and I see only one bug: fiveway up/down don't work on the emulator in landscape mode. But in portrait mode they still work.

kai771 commented:
I think they're replaced by left-right in landscape. Please test, as I can't do it now.

tigran123 commented:
Yes, you are right. It's ok because on the Kindle the keyboard is rotated too :)
(and on the emulator we'll just have to remember about this)

tigran123 commented:
Ok, in that case I can't break it on the emulator. I'll test it on the Kindle after breakfast...

Thank you very much --- very impressive work in such small number of lines of code, that's what I like!

kai771 commented:
I like it that way too :). But it wasn't me - it WAS simple. It just looked complicated. :)

tigran123 commented:
There is one inconsistency: if you switch Amazon framework to landscape mode then the filechooser will start in landscape mode, but after opening/closing a document it will be put into portrait mode. So, we have to choose between the following two options:
1. Force portrait mode in filechooser on startup as well (not just on closing the document).
2. Upon closing the document force not the portrait mode but whatever mode was used at startup. But this option implies the extra work of fixing filechooser to work in landscape mode upon closing the document (as we know it currently does not).

The minimalistic approach is option 1. and I vote for it. There is no real benefit in running filechooser in landscape mode as we all agreed, I think.

kai771 commented:
I agree with you. I'll fix it when I get home (using option 1). Or, you can merge now if you like, and submit the fix yourself. :)

tigran123 commented:
Just calling Screen:setRotationMode(0) before the call to FileChooser:choose() would not work, so I leave it up to you to fix it properly.

kai771 commented:
Ok, I'll take a look when I get back home. Won't be before evening though.

tigran123 commented:
Ok.

kai771 committed:
force portrait mode on KPV start too

kai771 commented:
There, that should do it... it even restores the mode that was active in the framework on close :) (not my doing, it just turned out that way :))
